src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs

```diff
@@ -0,0 +1,490 @@
+                if (runtimeException != null)
+                {
+                    TraceWriter.Verbose(string.Format(
+                        "Function runtime exception: {0}: {1}{2}{3}",
+                        runtimeException.ErrorRecord.InvocationInfo.InvocationName,
+                        runtimeException.Message,
+                        Environment.NewLine,
+                        runtimeException.ErrorRecord.InvocationInfo.PositionMessage));
+                }
+
+                TraceWriter.Verbose(string.Format("PowerShell Function completed (Failure, Id={0})", invocationId));
+                throw;
+            }
+            finally
+            {
+                _metrics.EndEvent(startedEvent);
+                DeleteProcessEnvironmentVariables(_environmentVariables);
+            }
+        }
+
+        protected override void OnScriptFileChanged(object sender, FileSystemEventArgs e)
+        {
+            if (_script == null)
+            {
+                // we're not loaded yet, so nothing to reload
+                return;
+            }
+
+            // The ScriptHost is already monitoring for changes to function.json, so we skip those
+            string fileExtension = Path.GetExtension(e.Name);
+            if (WatchedFileTypes.Contains(fileExtension))
+            {
+                _reloadScript();
+            }
+        }
+        // private void InvokePowerShellScript()
+        private void InvokePowerShellScript()
+        {
+            InitialSessionState iss = InitialSessionState.CreateDefault();
+            _moduleFiles = GetModuleFilePaths();
+
+            using (Runspace runspace = RunspaceFactory.CreateRunspace(iss))
+            {
+                runspace.Open();
+                RunspaceInvoke runSpaceInvoker = new RunspaceInvoke(runspace);
+                runSpaceInvoker.Invoke("Set-ExecutionPolicy -Scope Process -ExecutionPolicy Unrestricted");
+
+                using (
+                    System.Management.Automation.PowerShell powerShellInstance =
+                        System.Management.Automation.PowerShell.Create())
+                {
+                    powerShellInstance.Runspace = runspace;
+                    _moduleFiles = GetModuleFilePaths();
+                    if (_moduleFiles.Any())
+                    {
+                        powerShellInstance.AddCommand("Import-Module").AddArgument(_moduleFiles);
+                        LogLoadedModules();
+                    }
+
+                    _script = GetScript();
+                    powerShellInstance.AddScript(_script, true);
+
+                    PSDataCollection<PSObject> outputCollection = new PSDataCollection<PSObject>();
+                    outputCollection.DataAdded += OutputCollectionDataAdded;
+
+                    powerShellInstance.Streams.Error.DataAdded += ErrorDataAdded;
+
+                    IAsyncResult result = powerShellInstance.BeginInvoke<PSObject, PSObject>(null, outputCollection);
```

Comment:
We should pass a callback to BeginInvoke so we can:

1. Properly call EndInvoke
2. Eliminate the loop in line 178 and just await for the completion instead of calling sleep